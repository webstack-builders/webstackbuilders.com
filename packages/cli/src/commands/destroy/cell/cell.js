import { files as cellFiles } from '../../generate/cell/cell'
import { createYargsForComponentDestroy } from '../helpers'

export const { command, description, builder, handler, tasks } =
  createYargsForComponentDestroy({
    componentName: 'cell',
    filesFn: cellFiles,
  })
