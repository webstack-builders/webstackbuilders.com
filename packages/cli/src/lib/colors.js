import chalk from 'chalk'

/**
 * To keep a consistent color/style palette between cli packages, such as
 * @redwood/cli and @redwood/create-redwood-app, please keep them compatible
 * with one and another. We'll might split up and refactor these into a
 * separate package when there is a strong motivation behind it.
 *
 * Current files:
 *
 * - packages/cli/src/lib/colors.js (this file)
 * - packages/create-redwood-app/src/create-redwood-app.js
 */
export default {
  error: chalk.bold.red,
  warning: chalk.keyword('orange'),
  green: chalk.green,
  info: chalk.grey,
  bold: chalk.bold,
  underline: chalk.underline,
}
