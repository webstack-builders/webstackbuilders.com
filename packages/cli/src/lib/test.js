/* eslint-env jest */

// Include at the top of your tests. Automatically mocks out the file system
//
// import { loadComponentFixture } from 'src/lib/test'
//
// test('true is true', () => {
//   expect('some output').toEqual(loadComponentFixture('component', 'filename.js'))
// })

import fs from 'fs'
import path from 'path'

jest.mock('@redwoodjs/internal', () => {
  const path = require('path')
  return {
    ...jest.requireActual('@redwoodjs/internal'),
    generate: () => {},
    getPaths: () => {
      const BASE_PATH = '/path/to/project'
      return {
        base: BASE_PATH,
        api: {
          dataMigrations: path.join(BASE_PATH, './api/prisma/dataMigrations'),
          db: path.join(global.__dirname, 'fixtures'), // this folder
          src: path.join(BASE_PATH, './api/src'),
          services: path.join(BASE_PATH, './api/src/services'),
          graphql: path.join(BASE_PATH, './api/src/graphql'),
          functions: path.join(BASE_PATH, './api/src/functions'),
        },
        web: {
          src: path.join(BASE_PATH, './web/src'),
          routes: path.join(BASE_PATH, 'web/src/Routes.js'),
          components: path.join(BASE_PATH, '/web/src/components'),
          layouts: path.join(BASE_PATH, '/web/src/layouts'),
          pages: path.join(BASE_PATH, '/web/src/pages'),
        },
        scripts: path.join(BASE_PATH, 'scripts'),
        generated: {
          base: path.join(BASE_PATH, '.redwood'),
          schema: path.join(BASE_PATH, '.redwood/schema.graphql'),
          types: {
            includes: path.join(BASE_PATH, '.redwood/types/includes'),
            mirror: path.join(BASE_PATH, '.redwood/types/mirror'),
          },
        },
      }
    },
  }
})

global.__prettierPath = path.resolve(
  __dirname,
  './__tests__/fixtures/prettier.config.js'
)

jest.mock('path', () => {
  const path = jest.requireActual('path')
  return {
    ...path,
    join: (...paths) => {
      if (
        paths &&
        paths[0] === '/path/to/project' &&
        paths[1] === 'prettier.config.js'
      ) {
        return global.__prettierPath
      }
      return path.join(...paths)
    },
  }
})

export const generatorsRootPath = path.join(
  __dirname,
  '..',
  'commands',
  'generate'
)

/**
 * Loads the fixture for a generator by assuming a lot of the path structure
 * automatically:
 *
 *   `loadGeneratorFixture('scaffold', 'NamePage.js')`
 *
 * will return the contents of:
 *
 *   `cli/src/commands/generate/scaffold/__tests__/fixtures/NamePage.js`
 */
export const loadGeneratorFixture = (generator, name) => {
  return loadFixture(
    path.join(
      __dirname,
      '..',
      'commands',
      'generate',
      generator,
      '__tests__',
      'fixtures',
      name
    )
  )
}

/**
 * Returns the contents of a text file in a `fixtures` directory
 */
export const loadFixture = (filepath) => {
  return fs.readFileSync(filepath).toString()
}
