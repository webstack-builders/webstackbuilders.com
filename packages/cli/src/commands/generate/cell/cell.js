import pascalcase from 'pascalcase'
import pluralize from 'pluralize'

import { generate } from '@redwoodjs/internal'

import { transformTSToJS } from '../../../lib'
import { getSchema } from '../../../lib'
import { yargsDefaults } from '../../generate'
import {
  templateForComponentFile,
  createYargsForComponentGeneration,
  forcePluralizeWord,
  isWordNonPluralizable,
} from '../helpers'

const COMPONENT_SUFFIX = 'Cell'
const REDWOOD_WEB_PATH_NAME = 'components'

const getCellOperationNames = async () => {
  const { getProject } = await import('@redwoodjs/structure')

  return getProject()
    .cells.map((x) => {
      return x.queryOperationName
    })
    .filter(Boolean)
}

const uniqueOperationName = async (name, { index = 1, list = false }) => {
  let operationName = pascalcase(
    index <= 1 ? `find_${name}_query` : `find_${name}_query_${index}`
  )

  if (list) {
    operationName =
      index <= 1
        ? `${pascalcase(name)}Query`
        : `${pascalcase(name)}Query_${index}`
  }

  const cellOperationNames = await getCellOperationNames()
  if (!cellOperationNames.includes(operationName)) {
    return operationName
  }
  return uniqueOperationName(name, { index: index + 1 })
}

const getIdType = (model) => {
  return model.fields.find((field) => field.isId)?.type
}

export const files = async ({
  name,
  typescript: generateTypescript,
  ...options
}) => {
  let cellName = name
  let idType,
    model = null
  let templateNameSuffix = ''

  // Create a unique operation name.

  const shouldGenerateList =
    (isWordNonPluralizable(name) ? options.list : pluralize.isPlural(name)) ||
    options.list

  if (shouldGenerateList) {
    cellName = forcePluralizeWord(name)
    templateNameSuffix = 'List'
    // override operationName so that its find_operationName
  } else {
    // needed for the singular cell GQL query find by id case
    try {
      model = await getSchema(pascalcase(pluralize.singular(name)))
      idType = getIdType(model)
    } catch {
      // eat error so that the destroy cell generator doesn't raise when try to find prisma query engine in test runs
      // assume id will be Int, otherwise generated will keep throwing
      idType = 'Int'
    }
  }

  const operationName = await uniqueOperationName(cellName, {
    list: shouldGenerateList,
  })

  const cellFile = templateForComponentFile({
    name: cellName,
    suffix: COMPONENT_SUFFIX,
    extension: generateTypescript ? '.tsx' : '.js',
    webPathSection: REDWOOD_WEB_PATH_NAME,
    generator: 'cell',
    templatePath: `cell${templateNameSuffix}.tsx.template`,
    templateVars: {
      operationName,
      idType,
    },
  })

  const testFile = templateForComponentFile({
    name: cellName,
    suffix: COMPONENT_SUFFIX,
    extension: generateTypescript ? '.test.tsx' : '.test.js',
    webPathSection: REDWOOD_WEB_PATH_NAME,
    generator: 'cell',
    templatePath: 'test.js.template',
  })

  const storiesFile = templateForComponentFile({
    name: cellName,
    suffix: COMPONENT_SUFFIX,
    extension: generateTypescript ? '.stories.tsx' : '.stories.js',
    webPathSection: REDWOOD_WEB_PATH_NAME,
    generator: 'cell',
    templatePath: 'stories.js.template',
  })

  const mockFile = templateForComponentFile({
    name: cellName,
    suffix: COMPONENT_SUFFIX,
    extension: generateTypescript ? '.mock.ts' : '.mock.js',
    webPathSection: REDWOOD_WEB_PATH_NAME,
    generator: 'cell',
    templatePath: `mock${templateNameSuffix}.js.template`,
  })

  const files = [cellFile]

  if (options.stories) {
    files.push(storiesFile)
  }

  if (options.tests) {
    files.push(testFile)
  }

  if (options.stories || options.tests) {
    files.push(mockFile)
  }

  // Returns
  // {
  //    "path/to/fileA": "<<<template>>>",
  //    "path/to/fileB": "<<<template>>>",
  // }
  return files.reduce((acc, [outputPath, content]) => {
    const template = generateTypescript
      ? content
      : transformTSToJS(outputPath, content)

    return {
      [outputPath]: template,
      ...acc,
    }
  }, {})
}

export const { command, description, builder, handler } =
  createYargsForComponentGeneration({
    componentName: 'cell',
    filesFn: files,
    optionsObj: {
      ...yargsDefaults,
      list: {
        alias: 'l',
        default: false,
        description:
          'Use when you want to generate a cell for a list of the model name.',
        type: 'boolean',
      },
    },
    includeAdditionalTasks: () => {
      return [
        {
          title: `Generating types ...`,
          task: async () => {
            return generate()
          },
        },
      ]
    },
  })
