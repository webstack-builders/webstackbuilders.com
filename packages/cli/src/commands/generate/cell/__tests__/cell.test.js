global.__dirname = __dirname
import path from 'path'

// Load mocks
import '../../../../lib/test'
import * as cell from '../cell'

jest.mock('@redwoodjs/structure', () => {
  return {
    getProject: () => ({
      cells: [{ queryOperationName: undefined }],
    }),
  }
})

let singleWordFiles,
  multiWordFiles,
  snakeCaseWordFiles,
  kebabCaseWordFiles,
  camelCaseWordFiles,
  withoutTestFiles,
  withoutStoryFiles,
  withoutTestAndStoryFiles,
  listFlagPassedIn,
  listInferredFromName,
  modelPluralMatchesSingularWithList,
  modelPluralMatchesSingularWithoutList

beforeAll(async () => {
  singleWordFiles = await cell.files({
    name: 'User',
    tests: true,
    stories: true,
    list: false,
  })
  multiWordFiles = await cell.files({
    name: 'UserProfile',
    tests: true,
    stories: true,
    list: false,
  })
  snakeCaseWordFiles = await cell.files({
    name: 'user_profile',
    tests: true,
    stories: true,
    list: false,
  })
  kebabCaseWordFiles = await cell.files({
    name: 'user-profile',
    tests: true,
    stories: true,
    list: false,
  })
  camelCaseWordFiles = await cell.files({
    name: 'userProfile',
    tests: true,
    stories: true,
    list: false,
  })
  withoutTestFiles = await cell.files({
    name: 'User',
    tests: false,
    stories: true,
    list: false,
  })
  withoutStoryFiles = await cell.files({
    name: 'User',
    tests: true,
    stories: false,
    list: false,
  })
  withoutTestAndStoryFiles = await cell.files({
    name: 'User',
    tests: false,
    stories: false,
    list: false,
  })

  listFlagPassedIn = await cell.files({
    name: 'Member',
    tests: true,
    stories: true,
    list: true,
  })

  listInferredFromName = await cell.files({
    name: 'Members',
    tests: true,
    stories: true,
  })

  modelPluralMatchesSingularWithList = await cell.files({
    name: 'equipment',
    tests: true,
    stories: true,
    list: true,
  })

  modelPluralMatchesSingularWithoutList = await cell.files({
    name: 'equipment',
    tests: true,
    stories: true,
    list: false,
  })
})

// Single Word Scenario: User
test('returns exactly 4 files', () => {
  expect(Object.keys(singleWordFiles).length).toEqual(4)
})

test('creates a cell component with a single word name', () => {
  expect(
    singleWordFiles[
      path.normalize('/path/to/project/web/src/components/UserCell/UserCell.js')
    ]
  ).toMatchSnapshot()
})

test('creates a cell test with a single word name', () => {
  expect(
    singleWordFiles[
      path.normalize(
        '/path/to/project/web/src/components/UserCell/UserCell.test.js'
      )
    ]
  ).toMatchSnapshot()
})

test('creates a cell stories with a single word name', () => {
  expect(
    singleWordFiles[
      path.normalize(
        '/path/to/project/web/src/components/UserCell/UserCell.stories.js'
      )
    ]
  ).toMatchSnapshot()
})

test('creates a cell mock with a single word name', () => {
  expect(
    singleWordFiles[
      path.normalize(
        '/path/to/project/web/src/components/UserCell/UserCell.mock.js'
      )
    ]
  ).toMatchSnapshot()
})

// Multi Word Scenario: UserProfile
test('creates a cell component with a multi word name', () => {
  expect(
    multiWordFiles[
      path.normalize(
        '/path/to/project/web/src/components/UserProfileCell/UserProfileCell.js'
      )
    ]
  ).toMatchSnapshot()
})

test('creates a cell test with a multi word name', () => {
  expect(
    multiWordFiles[
      path.normalize(
        '/path/to/project/web/src/components/UserProfileCell/UserProfileCell.test.js'
      )
    ]
  ).toMatchSnapshot()
})

test('creates a cell stories with a multi word name', () => {
  expect(
    multiWordFiles[
      path.normalize(
        '/path/to/project/web/src/components/UserProfileCell/UserProfileCell.stories.js'
      )
    ]
  ).toMatchSnapshot()
})

test('creates a cell mock with a multi word name', () => {
  expect(
    multiWordFiles[
      path.normalize(
        '/path/to/project/web/src/components/UserProfileCell/UserProfileCell.mock.js'
      )
    ]
  ).toMatchSnapshot()
})

// SnakeCase Word Scenario: user_profile
test('creates a cell component with a snakeCase word name', () => {
  expect(
    snakeCaseWordFiles[
      path.normalize(
        '/path/to/project/web/src/components/UserProfileCell/UserProfileCell.js'
      )
    ]
  ).toMatchSnapshot()
})

test('creates a cell test with a snakeCase word name', () => {
  expect(
    snakeCaseWordFiles[
      path.normalize(
        '/path/to/project/web/src/components/UserProfileCell/UserProfileCell.test.js'
      )
    ]
  ).toMatchSnapshot()
})

test('creates a cell stories with a snakeCase word name', () => {
  expect(
    snakeCaseWordFiles[
      path.normalize(
        '/path/to/project/web/src/components/UserProfileCell/UserProfileCell.stories.js'
      )
    ]
  ).toMatchSnapshot()
})

test('creates a cell mock with a snakeCase word name', () => {
  expect(
    snakeCaseWordFiles[
      path.normalize(
        '/path/to/project/web/src/components/UserProfileCell/UserProfileCell.mock.js'
      )
    ]
  ).toMatchSnapshot()
})

// KebabCase Word Scenario: user-profile
test('creates a cell component with a kebabCase word name', () => {
  expect(
    kebabCaseWordFiles[
      path.normalize(
        '/path/to/project/web/src/components/UserProfileCell/UserProfileCell.js'
      )
    ]
  ).toMatchSnapshot()
})

test('creates a cell test with a kebabCase word name', () => {
  expect(
    kebabCaseWordFiles[
      path.normalize(
        '/path/to/project/web/src/components/UserProfileCell/UserProfileCell.test.js'
      )
    ]
  ).toMatchSnapshot()
})

test('creates a cell stories with a kebabCase word name', () => {
  expect(
    kebabCaseWordFiles[
      path.normalize(
        '/path/to/project/web/src/components/UserProfileCell/UserProfileCell.stories.js'
      )
    ]
  ).toMatchSnapshot()
})

test('creates a cell mock with a kebabCase word name', () => {
  expect(
    kebabCaseWordFiles[
      path.normalize(
        '/path/to/project/web/src/components/UserProfileCell/UserProfileCell.mock.js'
      )
    ]
  ).toMatchSnapshot()
})

// camelCase Word Scenario: user-profile
test('creates a cell component with a camelCase word name', () => {
  expect(
    camelCaseWordFiles[
      path.normalize(
        '/path/to/project/web/src/components/UserProfileCell/UserProfileCell.js'
      )
    ]
  ).toMatchSnapshot()
})

test('creates a cell test with a camelCase word name', () => {
  expect(
    camelCaseWordFiles[
      path.normalize(
        '/path/to/project/web/src/components/UserProfileCell/UserProfileCell.test.js'
      )
    ]
  ).toMatchSnapshot()
})

test('creates a cell stories with a camelCase word name', () => {
  expect(
    camelCaseWordFiles[
      path.normalize(
        '/path/to/project/web/src/components/UserProfileCell/UserProfileCell.stories.js'
      )
    ]
  ).toMatchSnapshot()
})

test('creates a cell mock with a camelCase word name', () => {
  expect(
    camelCaseWordFiles[
      path.normalize(
        '/path/to/project/web/src/components/UserProfileCell/UserProfileCell.mock.js'
      )
    ]
  ).toMatchSnapshot()
})

test("doesn't include test file when --tests is set to false", () => {
  expect(Object.keys(withoutTestFiles)).toEqual([
    path.normalize(
      '/path/to/project/web/src/components/UserCell/UserCell.mock.js'
    ),
    path.normalize(
      '/path/to/project/web/src/components/UserCell/UserCell.stories.js'
    ),
    path.normalize('/path/to/project/web/src/components/UserCell/UserCell.js'),
  ])
})

test("doesn't include storybook file when --stories is set to false", () => {
  expect(Object.keys(withoutStoryFiles)).toEqual([
    path.normalize(
      '/path/to/project/web/src/components/UserCell/UserCell.mock.js'
    ),
    path.normalize(
      '/path/to/project/web/src/components/UserCell/UserCell.test.js'
    ),
    path.normalize('/path/to/project/web/src/components/UserCell/UserCell.js'),
  ])
})

test("doesn't include storybook and test files when --stories and --tests is set to false", () => {
  expect(Object.keys(withoutTestAndStoryFiles)).toEqual([
    path.normalize('/path/to/project/web/src/components/UserCell/UserCell.js'),
  ])
})

test('generates list cells if list flag passed in', () => {
  const CELL_PATH = path.normalize(
    '/path/to/project/web/src/components/MembersCell/MembersCell.js'
  )

  const TEST_PATH = path.normalize(
    '/path/to/project/web/src/components/MembersCell/MembersCell.test.js'
  )

  const STORY_PATH = path.normalize(
    '/path/to/project/web/src/components/MembersCell/MembersCell.stories.js'
  )

  const MOCK_PATH = path.normalize(
    '/path/to/project/web/src/components/MembersCell/MembersCell.mock.js'
  )

  // Check the file names
  expect(Object.keys(listFlagPassedIn)).toEqual([
    MOCK_PATH,
    TEST_PATH,
    STORY_PATH,
    CELL_PATH,
  ])

  // Check the contents
  expect(listFlagPassedIn[CELL_PATH]).toMatchSnapshot()
})

test('generates list cells if name is plural', () => {
  const CELL_PATH = path.normalize(
    '/path/to/project/web/src/components/MembersCell/MembersCell.js'
  )

  const TEST_PATH = path.normalize(
    '/path/to/project/web/src/components/MembersCell/MembersCell.test.js'
  )

  const STORY_PATH = path.normalize(
    '/path/to/project/web/src/components/MembersCell/MembersCell.stories.js'
  )

  const MOCK_PATH = path.normalize(
    '/path/to/project/web/src/components/MembersCell/MembersCell.mock.js'
  )

  // Check the file names
  expect(Object.keys(listInferredFromName)).toEqual([
    MOCK_PATH,
    TEST_PATH,
    STORY_PATH,
    CELL_PATH,
  ])

  // Check the contents
  expect(listInferredFromName[CELL_PATH]).toMatchSnapshot()
})

test('"equipment" with list flag', () => {
  const CELL_PATH = path.normalize(
    '/path/to/project/web/src/components/EquipmentListCell/EquipmentListCell.js'
  )

  const TEST_PATH = path.normalize(
    '/path/to/project/web/src/components/EquipmentListCell/EquipmentListCell.test.js'
  )

  const STORY_PATH = path.normalize(
    '/path/to/project/web/src/components/EquipmentListCell/EquipmentListCell.stories.js'
  )

  const MOCK_PATH = path.normalize(
    '/path/to/project/web/src/components/EquipmentListCell/EquipmentListCell.mock.js'
  )

  // Check the file names
  expect(Object.keys(modelPluralMatchesSingularWithList)).toEqual([
    MOCK_PATH,
    TEST_PATH,
    STORY_PATH,
    CELL_PATH,
  ])

  // Check the contents
  expect(modelPluralMatchesSingularWithList[CELL_PATH]).toMatchSnapshot()
})

test('"equipment" withOUT list flag should find equipment by id', () => {
  const CELL_PATH = path.normalize(
    '/path/to/project/web/src/components/EquipmentCell/EquipmentCell.js'
  )

  const TEST_PATH = path.normalize(
    '/path/to/project/web/src/components/EquipmentCell/EquipmentCell.test.js'
  )

  const STORY_PATH = path.normalize(
    '/path/to/project/web/src/components/EquipmentCell/EquipmentCell.stories.js'
  )

  const MOCK_PATH = path.normalize(
    '/path/to/project/web/src/components/EquipmentCell/EquipmentCell.mock.js'
  )

  // Check the file names
  expect(Object.keys(modelPluralMatchesSingularWithoutList)).toEqual([
    MOCK_PATH,
    TEST_PATH,
    STORY_PATH,
    CELL_PATH,
  ])

  // Check the contents
  expect(modelPluralMatchesSingularWithoutList[CELL_PATH]).toMatchSnapshot()
})
