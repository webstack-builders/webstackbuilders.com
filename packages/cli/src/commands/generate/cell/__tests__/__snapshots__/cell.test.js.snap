// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`"equipment" with list flag 1`] = `
"export const QUERY = gql\`
  query EquipmentListQuery {
    equipmentList {
      id
    }
  }
\`

export const Loading = () => <div>Loading...</div>

export const Empty = () => <div>Empty</div>

export const Failure = ({ error }) => (
  <div style={{ color: 'red' }}>Error: {error.message}</div>
)

export const Success = ({ equipmentList }) => {
  return (
    <ul>
      {equipmentList.map((item) => {
        return <li key={item.id}>{JSON.stringify(item)}</li>
      })}
    </ul>
  )
}
"
`;

exports[`"equipment" withOUT list flag should find equipment by id 1`] = `
"export const QUERY = gql\`
  query FindEquipmentQuery($id: Int!) {
    equipment: equipment(id: $id) {
      id
    }
  }
\`

export const Loading = () => <div>Loading...</div>

export const Empty = () => <div>Empty</div>

export const Failure = ({ error }) => (
  <div style={{ color: 'red' }}>Error: {error.message}</div>
)

export const Success = ({ equipment }) => {
  return <div>{JSON.stringify(equipment)}</div>
}
"
`;

exports[`creates a cell component with a camelCase word name 1`] = `
"export const QUERY = gql\`
  query FindUserProfileQuery($id: Int!) {
    userProfile: userProfile(id: $id) {
      id
    }
  }
\`

export const Loading = () => <div>Loading...</div>

export const Empty = () => <div>Empty</div>

export const Failure = ({ error }) => (
  <div style={{ color: 'red' }}>Error: {error.message}</div>
)

export const Success = ({ userProfile }) => {
  return <div>{JSON.stringify(userProfile)}</div>
}
"
`;

exports[`creates a cell component with a kebabCase word name 1`] = `
"export const QUERY = gql\`
  query FindUserProfileQuery($id: Int!) {
    userProfile: userProfile(id: $id) {
      id
    }
  }
\`

export const Loading = () => <div>Loading...</div>

export const Empty = () => <div>Empty</div>

export const Failure = ({ error }) => (
  <div style={{ color: 'red' }}>Error: {error.message}</div>
)

export const Success = ({ userProfile }) => {
  return <div>{JSON.stringify(userProfile)}</div>
}
"
`;

exports[`creates a cell component with a multi word name 1`] = `
"export const QUERY = gql\`
  query FindUserProfileQuery($id: Int!) {
    userProfile: userProfile(id: $id) {
      id
    }
  }
\`

export const Loading = () => <div>Loading...</div>

export const Empty = () => <div>Empty</div>

export const Failure = ({ error }) => (
  <div style={{ color: 'red' }}>Error: {error.message}</div>
)

export const Success = ({ userProfile }) => {
  return <div>{JSON.stringify(userProfile)}</div>
}
"
`;

exports[`creates a cell component with a single word name 1`] = `
"export const QUERY = gql\`
  query FindUserQuery($id: Int!) {
    user: user(id: $id) {
      id
    }
  }
\`

export const Loading = () => <div>Loading...</div>

export const Empty = () => <div>Empty</div>

export const Failure = ({ error }) => (
  <div style={{ color: 'red' }}>Error: {error.message}</div>
)

export const Success = ({ user }) => {
  return <div>{JSON.stringify(user)}</div>
}
"
`;

exports[`creates a cell component with a snakeCase word name 1`] = `
"export const QUERY = gql\`
  query FindUserProfileQuery($id: Int!) {
    userProfile: userProfile(id: $id) {
      id
    }
  }
\`

export const Loading = () => <div>Loading...</div>

export const Empty = () => <div>Empty</div>

export const Failure = ({ error }) => (
  <div style={{ color: 'red' }}>Error: {error.message}</div>
)

export const Success = ({ userProfile }) => {
  return <div>{JSON.stringify(userProfile)}</div>
}
"
`;

exports[`creates a cell mock with a camelCase word name 1`] = `
"// Define your own mock data here:
export const standard = () => ({
  userProfile: {
    id: 42,
  },
})
"
`;

exports[`creates a cell mock with a kebabCase word name 1`] = `
"// Define your own mock data here:
export const standard = () => ({
  userProfile: {
    id: 42,
  },
})
"
`;

exports[`creates a cell mock with a multi word name 1`] = `
"// Define your own mock data here:
export const standard = () => ({
  userProfile: {
    id: 42,
  },
})
"
`;

exports[`creates a cell mock with a single word name 1`] = `
"// Define your own mock data here:
export const standard = () => ({
  user: {
    id: 42,
  },
})
"
`;

exports[`creates a cell mock with a snakeCase word name 1`] = `
"// Define your own mock data here:
export const standard = () => ({
  userProfile: {
    id: 42,
  },
})
"
`;

exports[`creates a cell stories with a camelCase word name 1`] = `
"import { Loading, Empty, Failure, Success } from './UserProfileCell'
import { standard } from './UserProfileCell.mock'

export const loading = () => {
  return Loading ? <Loading /> : null
}

export const empty = () => {
  return Empty ? <Empty /> : null
}

export const failure = () => {
  return Failure ? <Failure error={new Error('Oh no')} /> : null
}

export const success = () => {
  return Success ? <Success {...standard()} /> : null
}

export default { title: 'Cells/UserProfileCell' }
"
`;

exports[`creates a cell stories with a kebabCase word name 1`] = `
"import { Loading, Empty, Failure, Success } from './UserProfileCell'
import { standard } from './UserProfileCell.mock'

export const loading = () => {
  return Loading ? <Loading /> : null
}

export const empty = () => {
  return Empty ? <Empty /> : null
}

export const failure = () => {
  return Failure ? <Failure error={new Error('Oh no')} /> : null
}

export const success = () => {
  return Success ? <Success {...standard()} /> : null
}

export default { title: 'Cells/UserProfileCell' }
"
`;

exports[`creates a cell stories with a multi word name 1`] = `
"import { Loading, Empty, Failure, Success } from './UserProfileCell'
import { standard } from './UserProfileCell.mock'

export const loading = () => {
  return Loading ? <Loading /> : null
}

export const empty = () => {
  return Empty ? <Empty /> : null
}

export const failure = () => {
  return Failure ? <Failure error={new Error('Oh no')} /> : null
}

export const success = () => {
  return Success ? <Success {...standard()} /> : null
}

export default { title: 'Cells/UserProfileCell' }
"
`;

exports[`creates a cell stories with a single word name 1`] = `
"import { Loading, Empty, Failure, Success } from './UserCell'
import { standard } from './UserCell.mock'

export const loading = () => {
  return Loading ? <Loading /> : null
}

export const empty = () => {
  return Empty ? <Empty /> : null
}

export const failure = () => {
  return Failure ? <Failure error={new Error('Oh no')} /> : null
}

export const success = () => {
  return Success ? <Success {...standard()} /> : null
}

export default { title: 'Cells/UserCell' }
"
`;

exports[`creates a cell stories with a snakeCase word name 1`] = `
"import { Loading, Empty, Failure, Success } from './UserProfileCell'
import { standard } from './UserProfileCell.mock'

export const loading = () => {
  return Loading ? <Loading /> : null
}

export const empty = () => {
  return Empty ? <Empty /> : null
}

export const failure = () => {
  return Failure ? <Failure error={new Error('Oh no')} /> : null
}

export const success = () => {
  return Success ? <Success {...standard()} /> : null
}

export default { title: 'Cells/UserProfileCell' }
"
`;

exports[`creates a cell test with a camelCase word name 1`] = `
"import { render } from '@redwoodjs/testing/web'
import { Loading, Empty, Failure, Success } from './UserProfileCell'
import { standard } from './UserProfileCell.mock'

describe('UserProfileCell', () => {
  it('renders Loading successfully', () => {
    expect(() => {
      render(<Loading />)
    }).not.toThrow()
  })

  it('renders Empty successfully', async () => {
    expect(() => {
      render(<Empty />)
    }).not.toThrow()
  })

  it('renders Failure successfully', async () => {
    expect(() => {
      render(<Failure error={new Error('Oh no')} />)
    }).not.toThrow()
  })

  // When you're ready to test the actual output of your component render
  // you could test that, for example, certain text is present:
  //
  // 1. import { screen } from '@redwoodjs/testing/web'
  // 2. Add test: expect(screen.getByText('Hello, world')).toBeInTheDocument()

  it('renders Success successfully', async () => {
    expect(() => {
      render(<Success userProfile={standard().userProfile} />)
    }).not.toThrow()
  })
})
"
`;

exports[`creates a cell test with a kebabCase word name 1`] = `
"import { render } from '@redwoodjs/testing/web'
import { Loading, Empty, Failure, Success } from './UserProfileCell'
import { standard } from './UserProfileCell.mock'

describe('UserProfileCell', () => {
  it('renders Loading successfully', () => {
    expect(() => {
      render(<Loading />)
    }).not.toThrow()
  })

  it('renders Empty successfully', async () => {
    expect(() => {
      render(<Empty />)
    }).not.toThrow()
  })

  it('renders Failure successfully', async () => {
    expect(() => {
      render(<Failure error={new Error('Oh no')} />)
    }).not.toThrow()
  })

  // When you're ready to test the actual output of your component render
  // you could test that, for example, certain text is present:
  //
  // 1. import { screen } from '@redwoodjs/testing/web'
  // 2. Add test: expect(screen.getByText('Hello, world')).toBeInTheDocument()

  it('renders Success successfully', async () => {
    expect(() => {
      render(<Success userProfile={standard().userProfile} />)
    }).not.toThrow()
  })
})
"
`;

exports[`creates a cell test with a multi word name 1`] = `
"import { render } from '@redwoodjs/testing/web'
import { Loading, Empty, Failure, Success } from './UserProfileCell'
import { standard } from './UserProfileCell.mock'

describe('UserProfileCell', () => {
  it('renders Loading successfully', () => {
    expect(() => {
      render(<Loading />)
    }).not.toThrow()
  })

  it('renders Empty successfully', async () => {
    expect(() => {
      render(<Empty />)
    }).not.toThrow()
  })

  it('renders Failure successfully', async () => {
    expect(() => {
      render(<Failure error={new Error('Oh no')} />)
    }).not.toThrow()
  })

  // When you're ready to test the actual output of your component render
  // you could test that, for example, certain text is present:
  //
  // 1. import { screen } from '@redwoodjs/testing/web'
  // 2. Add test: expect(screen.getByText('Hello, world')).toBeInTheDocument()

  it('renders Success successfully', async () => {
    expect(() => {
      render(<Success userProfile={standard().userProfile} />)
    }).not.toThrow()
  })
})
"
`;

exports[`creates a cell test with a single word name 1`] = `
"import { render } from '@redwoodjs/testing/web'
import { Loading, Empty, Failure, Success } from './UserCell'
import { standard } from './UserCell.mock'

describe('UserCell', () => {
  it('renders Loading successfully', () => {
    expect(() => {
      render(<Loading />)
    }).not.toThrow()
  })

  it('renders Empty successfully', async () => {
    expect(() => {
      render(<Empty />)
    }).not.toThrow()
  })

  it('renders Failure successfully', async () => {
    expect(() => {
      render(<Failure error={new Error('Oh no')} />)
    }).not.toThrow()
  })

  // When you're ready to test the actual output of your component render
  // you could test that, for example, certain text is present:
  //
  // 1. import { screen } from '@redwoodjs/testing/web'
  // 2. Add test: expect(screen.getByText('Hello, world')).toBeInTheDocument()

  it('renders Success successfully', async () => {
    expect(() => {
      render(<Success user={standard().user} />)
    }).not.toThrow()
  })
})
"
`;

exports[`creates a cell test with a snakeCase word name 1`] = `
"import { render } from '@redwoodjs/testing/web'
import { Loading, Empty, Failure, Success } from './UserProfileCell'
import { standard } from './UserProfileCell.mock'

describe('UserProfileCell', () => {
  it('renders Loading successfully', () => {
    expect(() => {
      render(<Loading />)
    }).not.toThrow()
  })

  it('renders Empty successfully', async () => {
    expect(() => {
      render(<Empty />)
    }).not.toThrow()
  })

  it('renders Failure successfully', async () => {
    expect(() => {
      render(<Failure error={new Error('Oh no')} />)
    }).not.toThrow()
  })

  // When you're ready to test the actual output of your component render
  // you could test that, for example, certain text is present:
  //
  // 1. import { screen } from '@redwoodjs/testing/web'
  // 2. Add test: expect(screen.getByText('Hello, world')).toBeInTheDocument()

  it('renders Success successfully', async () => {
    expect(() => {
      render(<Success userProfile={standard().userProfile} />)
    }).not.toThrow()
  })
})
"
`;

exports[`generates list cells if list flag passed in 1`] = `
"export const QUERY = gql\`
  query MembersQuery {
    members {
      id
    }
  }
\`

export const Loading = () => <div>Loading...</div>

export const Empty = () => <div>Empty</div>

export const Failure = ({ error }) => (
  <div style={{ color: 'red' }}>Error: {error.message}</div>
)

export const Success = ({ members }) => {
  return (
    <ul>
      {members.map((item) => {
        return <li key={item.id}>{JSON.stringify(item)}</li>
      })}
    </ul>
  )
}
"
`;

exports[`generates list cells if name is plural 1`] = `
"export const QUERY = gql\`
  query MembersQuery {
    members {
      id
    }
  }
\`

export const Loading = () => <div>Loading...</div>

export const Empty = () => <div>Empty</div>

export const Failure = ({ error }) => (
  <div style={{ color: 'red' }}>Error: {error.message}</div>
)

export const Success = ({ members }) => {
  return (
    <ul>
      {members.map((item) => {
        return <li key={item.id}>{JSON.stringify(item)}</li>
      })}
    </ul>
  )
}
"
`;
